include/boost/geometry/algorithms/similarity.hpp

```diff
@@ -0,0 +1,330 @@
+// Boost.Geometry
+
+// Copyright (c) 2023 Barend Gehrels, Amsterdam, the Netherlands.
+
+// Use, modification and distribution is subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+
+#ifndef BOOST_GEOMETRY_ALGORITHMS_SIMILARITY_HPP
+#define BOOST_GEOMETRY_ALGORITHMS_SIMILARITY_HPP
+
+#include <cstdint>
+#include <vector>
+
+#include <boost/geometry/core/point_type.hpp>
+#include <boost/geometry/algorithms/distance.hpp>
+#include <boost/geometry/algorithms/length.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+template <typename CoordinateType>
+struct similarity_info
+{
+    CoordinateType distance = 0;
+    bool is_reversed = false;
+}; 
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace similarity
+{
+
+// Structure containing projections of a geometry on another geometry,
+// and additionally the points of the geometry itself
+template <typename Point>
+struct projection
+{
+    using coor_t = typename coordinate_type<Point>::type;
+
+    // Indicates if it is a projection of the other geometry, or a point of the own geometry.
+    bool is_projection = true;
+
+    // For a projection of P on Q, this is the index in P
+    // For a non projected point, it is also the index in P
+    std::size_t source_index = 0;
+
+    // For a projection of P on Q, this is the index of a segment in Q
+    // For a non projected point, it is identical to source_index
+    std::size_t segment_index = 0;
+
+    // For a projection of P on Q, this is the offset on the segment in Q
+    // For a non projected point, the offset is always 0.0
+    coor_t offset = 0;
+
+    // For a projection of P on Q, this is the distance from point P 
+    // to its projection in Q.
+    // For a non projected point, it is 0.0
+    coor_t distance = 0;
+
+    // The projected point, or the point of the geometry itself
+    Point point;
+};
+
+template <typename Point, typename Geometry>
+auto get_projection(std::size_t index, Point const& point, Geometry const& target)
+{
+    using coor_t = typename coordinate_type<Point>::type;
+    using closest_strategy = geometry::strategy::closest_points::detail::compute_closest_point_to_segment<coor_t>;
+    projection<Point> result;
+
+    for (std::size_t i = 0; i + 1 < target.size(); i++)
+    {
+        auto const& current = target[i];
+        auto const& next = target[i + 1];
+        projection<Point> other;
+        auto const pp = closest_strategy::apply(point, current, next);
```

Comment:
I looked at it. It seems that some code duplication is also there in the `detail::closest_feature::point_to_point_range` function, where the closest segment is determined based on (comparable) distance. Which, internally, already finds the projection - but it is discarded (in the strategy itself).
Then the closest point is calculated.

Here I do it the other way round, calling first closest point strategy (which already does some calculations to find it - but does not calculate the distance yet) and then calculate its distance (I could use comparable instead - changed it right now).

I'm not sure if it is worth the effort to harmonize this code for this reason only, it will make it more complex.

What could be useful is a strategy which gives all the internal information to avoid code duplication in other places. Something like that is already in extensions (`calculate_distance_info`) but never made it. So maybe also that is not really worth the effort...